File: 141-linked-list-cycle.cpp
// https://leetcode.com/problems/linked-list-cycle
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if(!head || !head->next)
            return false;
        ListNode* node1 = head;
        ListNode* node2 = head->next;

        while(node1 && node2){
            if(node1 == node2){
                return true;
            }
            node1 = node1->next;
            if(node2->next){
                node2 = node2->next->next;
            }else{
                return false;
            }
        }

        return false;
    }
};

// Extra testcases:
// []
// -1

File: 141-linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if not head or not head.next:
            return False
        fast = head.next.next

        while fast:
            head = head.next
            fast = fast.next
            if fast:
                fast = fast.next
            else:
                return False
            if fast == head:
                return True

        return False
            
        
